experiments/202005-bazel-rules-typescript/boat/src/boat.ts

```diff
@@ -0,0 +1,10 @@
+import "../../shipping-container-types/types"
```

Comment:
is there a way to import relative to root, instead of to the file?

Reply:
Yeah another good question.

- yes, you can, using "paths" in tsconfig.json. I used to use this, but it was related to my effort to achieve module independence. It's also arguably kind of nice - imports become just stuff like `import "shipping-container-types`
- now that I'm using more modern typescript, and the entire topic of "how to I parallelize/optimize ts compilation" has a different and better story, I'm wondering if we really want to use this module indirection...
- fun fact: in my usage of "paths", Intellij has some corner-case confusion around importing that's highly annoying. In one case, if a ts source file is sibling to another ts file, the compiler doesn't (or didn't, in my setup) like it when you use module-style imports for those. So I had to recognize this and correct it manually. I had a list of about 3 common issues I used to watch out for re: Intellij's automatic ts import capability.
- ...so I left this as relative importing. The upside of relative importing is that it could not be easier to immediately understand where the code in question lives. One downside used to be performance - that's no longer the case. Another downside that still exists is that you have specific paths to things strewn all over the place. OTOH, when you control the code, refactoring tools make this easily manageable.

Reply:
> ...so I left this as relative importing. The upside of relative importing is that it could not be easier to immediately understand where the code in question lives. One downside used to be performance - that's no longer the case. Another downside that still exists is that you have specific paths to things strewn all over the place. OTOH, when you control the code, refactoring tools make this easily manageable.

The thing I'm worried about is now in intellij you can only have one module src/, because you can't do relative imports across intellij modules and have intellij still able to build it.

Maybe seperate experiment, but we should rig it up and see what intellij does.

Were you picturing one intellij module with a src/ directory and nested build files for bazel, or multiple top level intellij modules with dependencies, kindof like program.world java is?